File: android/app/src/main/java/com/coinswallet/coins/BitcoinCoin.java
// BitcoinCoin.java
package com.coinswallet.coins;

import com.coinswallet.CoinFactory;
import com.coinswallet.Utils;


import wallet.core.jni.Derivation;
import wallet.core.jni.HDVersion;
import wallet.core.jni.HDWallet;
import wallet.core.jni.CoinType;
import wallet.core.jni.Purpose;


public class BitcoinCoin extends CoinFactory.Coin {
    private final HDWallet wallet;

    public BitcoinCoin(String mnemonic) {
        super(mnemonic);
        this.wallet = super.wallet;
    }

    @Override
    public String getNewAddress(Boolean isTestNet) {
        // Derive a new private key for each call
        Derivation derivation = Derivation.BITCOINSEGWIT;
        if (isTestNet) {
            derivation = Derivation.BITCOINTESTNET;
        }
        return wallet.getAddressDerivation(CoinType.BITCOIN, derivation);
    }

    @Override
    public String getPrivateKey(Boolean isTestNet) {
        Derivation derivation = Derivation.BITCOINSEGWIT;
        if (isTestNet) {
            derivation = Derivation.BITCOINTESTNET;
        }
        byte[] privateKeyBytes = wallet.getKeyDerivation(CoinType.BITCOIN,derivation).data();
        byte[] prefix = new byte[]{(byte) 0x80};
        byte[] testnetPrefix = new byte[]{(byte) 0xef};
        return Utils.convertPrivateKeytoWIF(privateKeyBytes,isTestNet,prefix,testnetPrefix);
    }

    @Override
    public String getExtendedPublicKey(Boolean isTestNet) {
        Derivation derivation = Derivation.BITCOINSEGWIT;
        HDVersion version = HDVersion.ZPUB;
        if(isTestNet){
            derivation =  Derivation.BITCOINTESTNET;
        }
        return wallet.getExtendedPublicKeyDerivation(Purpose.BIP84,CoinType.BITCOIN,derivation,version);
    }

    @Override
    public String getExtendedPrivateKey(Boolean isTestNet) {
        Derivation derivation = Derivation.BITCOINSEGWIT;
        HDVersion version = HDVersion.ZPRV;
        if(isTestNet){
            derivation =  Derivation.BITCOINTESTNET;
        }
        return wallet.getExtendedPrivateKeyDerivation(Purpose.BIP84,CoinType.BITCOIN,derivation,version);
    }


    @Override
    public String signTransaction(String rawData) {
        // You need to implement this method according to the specific requirements of
        // Bitcoin transactions.
        return null;
    }
}

File: ios/BitcoinCoin.swift
//
//  BitcoinCoin.swift
//  coinswallet
//
//  Created by Tal Grynbaum on 30/06/2023.
//


import Foundation
import WalletCore

class BitcoinCoin: CoinFactory.Coin {
  private var addressIndex: Int = 0
  
  override init(mnemonic: String) {
    super.init(mnemonic: mnemonic)
  }
  
  override func getNewAddress(isTestNet:Bool) -> String {
    var derivation = Derivation.bitcoinSegwit;
    if(isTestNet){
      derivation =  Derivation.bitcoinTestnet
    }
    let address  = wallet.getAddressDerivation(coin: .bitcoin, derivation: derivation)
    addressIndex += 1
    return address
  }
  
  override func getExtendedPublicKey(isTestNet:Bool) -> String {
    var derivation = Derivation.bitcoinSegwit;
    var version = HDVersion.zpub;
    if(isTestNet){
      derivation =  Derivation.bitcoinTestnet
      version = HDVersion.vpub
    }
    let extendedKey = wallet.getExtendedPublicKeyDerivation(purpose: .bip84, coin: .bitcoin, derivation: derivation, version: version);
    return extendedKey
  }
  
  override func getExtendedPrivateKey(isTestNet:Bool) -> String {
    var derivation = Derivation.bitcoinSegwit;
    var version = HDVersion.zprv;
    if(isTestNet){
      derivation =  Derivation.bitcoinTestnet
      version = HDVersion.vprv
    }
    let extendedKey = wallet.getExtendedPrivateKeyDerivation(purpose: .bip84, coin: .bitcoin, derivation: derivation, version: version);
    return extendedKey
  }
  
  override func getPrivateKey(isTestNet:Bool) -> String {
    
    var derivation = Derivation.bitcoinSegwit;
    if(isTestNet){
      derivation =  Derivation.bitcoinTestnet
    }
    let privateKeyBytes = wallet.getKeyDerivation(coin: .bitcoin, derivation: derivation).data
    return Utils.convertToWif(data: privateKeyBytes, isTestNet: isTestNet, prefix: [0x80], testNetPrefix: [0xef])
  }
  
  override func signTransaction(rawData: String) -> String {
    // Implement the logic to sign a Bitcoin transaction using the provided raw data
    return ""
  }
}
